src/arfsdao.ts

```diff
@@ -779,3 +1345,572 @@ export class ArFSPrivateDriveBuilder {
+	}
+}
+
+export interface ArFSWithPath {
+	readonly path: string;
+	readonly txIdPath: string;
+	readonly entityIdPath: string;
+}
+
+// TODO: replace all 'FileOrFolder' ocurrencies with 'Directory' (?)
+export class ArFSPublicFileOrFolderData extends ArFSFileOrFolderEntity implements ArFSWithPath {
+	constructor(
+		appName: string,
+		appVersion: string,
+		arFS: string,
+		contentType: string,
+		driveId: string,
+		entityType: string,
+		name: string,
+		txId: string,
+		unixTime: number,
+		parentFolderId: string,
+		readonly entityId: string,
+		readonly path: string,
+		readonly txIdPath: string,
+		readonly entityIdPath: string
+	) {
+		super(
+			appName,
+			appVersion,
+			arFS,
+			contentType,
+			driveId,
+			entityType,
+			name,
+			txId,
+			unixTime,
+			parentFolderId,
+			entityId
+		);
+	}
+}
+
+export class ArFSPrivateFileOrFolderData extends ArFSFileOrFolderEntity implements ArFSWithPath {
+	constructor(
+		appName: string,
+		appVersion: string,
+		arFS: string,
+		contentType: ContentType,
+		driveId: DriveID,
+		entityType: EntityType,
+		name: string,
+		txId: TransactionID,
+		unixTime: number,
+		parentFolderId: FolderID,
+		readonly entityId: string,
+		readonly cipher: string,
+		readonly cipherIV: CipherIV,
+		readonly path: string,
+		readonly txIdPath: string,
+		readonly entityIdPath: string
+	) {
+		super(
+			appName,
+			appVersion,
+			arFS,
+			contentType,
+			driveId,
+			entityType,
+			name,
+			txId,
+			unixTime,
+			parentFolderId,
+			entityId
+		);
+	}
+}
+
+export class ArFSPublicFile extends ArFSFileOrFolderEntity {
+	constructor(
+		readonly appName: string,
+		readonly appVersion: string,
+		readonly arFS: string,
+		readonly contentType: ContentType,
+		readonly driveId: DriveID,
+		readonly entityType: EntityType,
+		readonly name: string,
+		readonly txId: string,
+		readonly unixTime: number,
+		readonly parentFolderId: FolderID,
+		readonly fileId: FileID
+	) {
+		super(
+			appName,
+			appVersion,
+			arFS,
+			contentType,
+			driveId,
+			entityType,
+			name,
+			txId,
+			unixTime,
+			parentFolderId,
+			fileId
+		);
+	}
+}
+
+export class ArFSPublicFileBuilder {
+	appName?: string;
+	appVersion?: string;
+	arFS?: string;
+	contentType?: ContentType;
+	driveId?: DriveID;
+	entityType?: EntityType;
+	name?: string;
+	txId?: TransactionID;
+	unixTime?: number;
+	parentFolderId?: string;
+	fileId?: string;
+
+	build(): ArFSPublicFile {
+		if (
+			this.appName?.length &&
+			this.appVersion?.length &&
+			this.arFS?.length &&
+			this.contentType?.length &&
+			this.driveId?.length &&
+			this.entityType?.length &&
+			this.name?.length &&
+			this.txId?.length &&
+			this.unixTime &&
+			this.fileId?.length
+		) {
+			return new ArFSPublicFile(
+				this.appName,
+				this.appVersion,
+				this.arFS,
+				this.contentType,
+				this.driveId,
+				this.entityType,
+				this.name,
+				this.txId,
+				this.unixTime,
+				this.parentFolderId || 'root folder',
```

Comment:
There should always be a parent folder for files.

Reply:
There: 167564645ac7cd3809605c785e722b7a67078a84